# 46장 제너레이터와 async/await
## 46.1 제너레이터란?
ES6에서 도입된 제너레이터는 코드 블록의 실행을 일시 중지 했다가 필요한 시점에 재개할 수 있는 특수한 함수
*일반함수와 제너레이터의 차이*
1. 제너레이터 함수는 함수 호출자에게 함수 실행의 제어권을 양도 할 수 있음
	* 일반함수 : 제어권이 함수에게 넘어가고, 함수 코드를 일괄 실행
	* 제너테이터 : 함수 호출자에게 제어권을 양도(yield)할 수 있음
2. 제너레이터 함수는 함수 호출자와 함수의 상태를 주고받을 수 있음
	* 일반 함수 : 매개변수를 통해 함수 외부에서 값을 주입 받고, 함수 코드를 일괄 실행하여 결과 값을 함수 외부로 반환 (외부에서 함수 내부로 값을 전달하여 함수의 상태 변경 불가능)
	* 제너레이터 : 제너레이터 함수는 함수 호출자와 양방향으로 함수의 상태를 주고받을 수 있음 (호출자에게 상태를 전달할 수 있고, 함수 호출자로부터 상태를 전달 받을 수 있음) 
3. 제너레이터 함수를 호출하면 제너레이터 객체를 반환함
	* 일반 함수 : 코드 일괄 실행 후 값을 반환
	* 제너레이터 : 함수 코드를 실행하는 것이 아니라, 이터러블이면서 동시에 제너레이터인 객체를 반환

## 제너레이터 함수의 정의
```javascript
// 제너레이터 함수 선언문
function* genDecFunc() {
  yield 1;
}

// 제너레이터 함수 표현식
const genExpFunc = function* () {
  yield 1;
};

// 제너레이터 메서드
const obj = {
  * genObjMethod() {
    yield 1;
  }
};

// 제너레이터 클래스
class MyClass {
  * genClsMethod() {
    yield 1;
  }
}
```
* `function*` 키워드로 선언
	* 위치는 function키워드와 함수 이름 사이 아무곳에나 위치 가능함
	* 일관성을 위해 바로 function뒤에 붙이는 걸 추천
```js
function* genFunc() { yield 1; }

function * genFunc() { yield 1; }

function *genFunc() { yield 1; }

function*genFunc() { yield 1; }
```
* 하나 이상의 `yield` 표현식을 포함
* 제너레이터 함수는 화살표 함수로 정의할 수 없음
* 제너레이터 함수는 new 연산자와 함께 생성자 함수로 호출할 수 없음

## 46.3 제너레이터 객체
제너레이터 함수를 호출하면 제너레이터 객체를 생성해 반환함
*제너레이터 객체*
* 이터러블 : `Symbol.iterator` 메서드를 상속
* 이터레이터 : `value`, `done` 프로퍼티와 `next` 메서드
	* `Symbol.iterator` 메서드를 호출해서 이터레이터를 생성할 필요가 없음
```javascript
// 제너레이터 함수
function* genFunc() {
  yield 1;
  yield 2;
  yield 3;
}

// 제너레이터 함수를 호출하면 제너레이터 객체를 반환한다.
const generator = genFunc();

// 제너레이터 객체는 이터러블이면서 동시에 이터레이터다.
// 이터러블은 Symbol.iterator 메서드를 직접 구현하거나 프로토타입 체인을 통해 상속받은 객체다.
console.log(Symbol.iterator in generator); // true
// 이터레이터는 next 메서드를 갖는다.
console.log('next' in generator); // true
```

* 제너레이터 객체는 이터레이터 이지만, 이터레이터에 없는 `return`,  `throw` 라는 메서드를 가짐

*제너레이터 메서드*
* next 
	* 제너레이터 함수의 `yield` 표현식까지 코드 블록을 실행
	* `value`  = yield된 값,  `done` = false 로된 이터레이터 result 객체를 반환
* return
	* `value`  = 인수로 전달 받은 값,  `done` = true 로된 이터레이터 result 객체를 반환
* throw
	* 인수로 전달받은 에러를 발생시키고,  `value`  = undefined,  `done` = true 로된 이터레이터 result 객체를 반환

```javascript
function* genFunc() {
  try {
    yield 1;
    yield 2;
    yield 3;
  } catch (e) {
    console.error(e);
  }
}

const generator1 = genFunc();

console.log(generator1.next()); // {value: 1, done: false}
console.log(generator1.return('End!')); // {value: "End!", done: true}

const generator2 = genFunc();

console.log(generator2.next()); // {value: 1, done: false}
console.log(generator2.throw('Error!')); // {value: undefined, done: true}
```
![image-20220109002638231](/Users/jojunhyeong/Workspace/deepdive_study/images/image-20220109002638231.png)

## 46.4 제너레이터의 일시 중지와 재개
제너레이터는 함수 호출자가 yield 키워드와 next 메서드를 통해 실행을 일시 중지했다가, 필요한 시점에 다시 재개할 수 있음
* 제너레이터 함수는 함수의 코드블록을 실행하는 것이 아니라, 제너레이터 객체를 반환함
* 이터레이터이기 때문에, next 메서드를 호출할 수 있고, next 메서드는 코드 블럭을 yield 까지만 실행함
* yield 키워드는 제너레이터의 함수의 실행을 일시 중지 시키거나, yield 뒤에 오는 표현식의 평가 결과를 제너레이터 함수 호출자에게 반환함
```javascript
// 제너레이터 함수
function* genFunc() {
  yield 1;
  yield 2;
  yield 3;
}

// 제너레이터 함수를 호출하면 제너레이터 객체를 반환한다.
// 이터러블이면서 동시에 이터레이터인 제너레이터 객체는 next 메서드를 갖는다.
const generator = genFunc();

// 처음 next 메서드를 호출하면 첫 번째 yield 표현식까지 실행되고 일시 중지된다.
// next 메서드는 이터레이터 리절트 객체({value, done})를 반환한다.
// value 프로퍼티에는 첫 번째 yield 표현식에서 yield된 값 1이 할당된다.
// done 프로퍼티에는 제너레이터 함수가 끝까지 실행되었는지를 나타내는 false가 할당된다.
console.log(generator.next()); // {value: 1, done: false}

// 다시 next 메서드를 호출하면 두 번째 yield 표현식까지 실행되고 일시 중지된다.
// next 메서드는 이터레이터 리절트 객체({value, done})를 반환한다.
// value 프로퍼티에는 두 번째 yield 표현식에서 yield된 값 2가 할당된다.
// done 프로퍼티에는 제너레이터 함수가 끝까지 실행되었는지를 나타내는 false가 할당된다.
console.log(generator.next()); // {value: 2, done: false}

// 다시 next 메서드를 호출하면 세 번째 yield 표현식까지 실행되고 일시 중지된다.
// next 메서드는 이터레이터 리절트 객체({value, done})를 반환한다.
// value 프로퍼티에는 세 번째 yield 표현식에서 yield된 값 3이 할당된다.
// done 프로퍼티에는 제너레이터 함수가 끝까지 실행되었는지를 나타내는 false가 할당된다.
console.log(generator.next()); // {value: 3, done: false}

// 다시 next 메서드를 호출하면 남은 yield 표현식이 없으므로 제너레이터 함수의 마지막까지 실행한다.
// next 메서드는 이터레이터 리절트 객체({value, done})를 반환한다.
// value 프로퍼티에는 제너레이터 함수의 반환값 undefined가 할당된다.
// done 프로퍼티에는 제너레이터 함수가 끝까지 실행되었음을 나타내는 true가 할당된다.
console.log(generator.next()); // {value: undefined, done: true}
```
* next
	* yield 까지 실행되고 일시 중지(suspend)
		* 함수의 제어권이 호출자로 양도(yield)됨
		* 다시 한번 더 호출하면, 일시 중지 되었던 지점부터, 다음 yield 를 만날때 까지 실행이 되고 다시 일시 중지가 되는 것을 반복함
	* 반환되는 값
		* 이터레이터 리절트 객체
			* value: yield 뒤에 나오는 표현식의 평가 값, 함수가 종료(return)시 반환 값이 할당
			* done : 함수가 종료(return) 되지 않았다면 false, 종료(return) 되었다면  true
	* 이터레이터의 next와 달리, 제너레이터의 next는 메서드에 인수를 전달할 수 있음
		* 인수를 전달하면, yield  표현식을 할당받는 변수에 할당됨
```javascript
function* genFunc() {
  // 처음 next 메서드를 호출하면 첫 번째 yield 표현식까지 실행되고 일시 중지된다.
  // 이때 yield된 값 1은 next 메서드가 반환한 이터레이터 리절트 객체의 value 프로퍼티에 할당된다.
  // x 변수에는 아직 아무것도 할당되지 않았다. x 변수의 값은 next 메서드가 두 번째 호출될 때 결정된다.
  const x = yield 1;

  // 두 번째 next 메서드를 호출할 때 전달한 인수 10은 첫 번째 yield 표현식을 할당받는 x 변수에 할당된다.
  // 즉, const x = yield 1;은 두 번째 next 메서드를 호출했을 때 완료된다.
  // 두 번째 next 메서드를 호출하면 두 번째 yield 표현식까지 실행되고 일시 중지된다.
  // 이때 yield된 값 x + 10은 next 메서드가 반환한 이터레이터 리절트 객체의 value 프로퍼티에 할당된다.
  const y = yield (x + 10);

  // 세 번째 next 메서드를 호출할 때 전달한 인수 20은 두 번째 yield 표현식을 할당받는 y 변수에 할당된다.
  // 즉, const y = yield (x + 10);는 세 번째 next 메서드를 호출했을 때 완료된다.
  // 세 번째 next 메서드를 호출하면 함수 끝까지 실행된다.
  // 이때 제너레이터 함수의 반환값 x + y는 next 메서드가 반환한 이터레이터 리절트 객체의 value 프로퍼티에 할당된다.
  // 일반적으로 제너레이터의 반환값은 의미가 없다.
  // 따라서 제너레이터에서는 값을 반환할 필요가 없고 return은 종료의 의미로만 사용해야 한다.
  return x + y;
}

// 제너레이터 함수를 호출하면 제너레이터 객체를 반환한다.
// 이터러블이며 동시에 이터레이터인 제너레이터 객체는 next 메서드를 갖는다.
const generator = genFunc(0);

// 처음 호출하는 next 메서드에는 인수를 전달하지 않는다.
// 만약 처음 호출하는 next 메서드에 인수를 전달하면 무시된다.
// next 메서드가 반환한 이터레이터 리절트 객체의 value 프로퍼티에는 첫 번째 yield된 값 1이 할당된다.
let res = generator.next();
console.log(res); // {value: 1, done: false}

// next 메서드에 인수로 전달한 10은 genFunc 함수의 x 변수에 할당된다.
// next 메서드가 반환한 이터레이터 리절트 객체의 value 프로퍼티에는 두 번째 yield된 값 20이 할당된다.
res = generator.next(10);
console.log(res); // {value: 20, done: false}

// next 메서드에 인수로 전달한 20은 genFunc 함수의 y 변수에 할당된다.
// next 메서드가 반환한 이터레이터 리절트 객체의 value 프로퍼티에는 제너레이터 함수의 반환값 30이 할당된다.
res = generator.next(20);
console.log(res); // {value: 30, done: true}
```

## 46.5 제너레이터의 활용
### 46.5.1 이터러블의 구현
제너레이터 함수를 사용하면 이터레이션 프로토콜을 준수해 이터러블을 생성하는 방식보다 간단히 이터러블을 구현할 수 있음
*무한 피보나치 수열*
* 이터레이션 프로토콜을 준수하는 방법
```javascript
// 무한 이터러블을 생성하는 함수
const infiniteFibonacci = (function () {
  let [pre, cur] = [0, 1];

  return {
    [Symbol.iterator]() { return this; },
    next() {
      [pre, cur] = [cur, pre + cur];
      // 무한 이터러블이므로 done 프로퍼티를 생략한다.
      return { value: cur };
    }
  };
}());

// infiniteFibonacci는 무한 이터러블이다.
for (const num of infiniteFibonacci) {
  if (num > 10000) break;
  console.log(num); // 1 2 3 5 8...2584 4181 6765
}
```
* 제너레이터를 이용해 구현한 방법
```javascript
// 무한 이터러블을 생성하는 제너레이터 함수
const infiniteFibonacci = (function* () {
  let [pre, cur] = [0, 1];

  while (true) {
    [pre, cur] = [cur, pre + cur];
    yield cur;
  }
}());

// infiniteFibonacci는 무한 이터러블이다.
for (const num of infiniteFibonacci) {
  if (num > 10000) break;
  console.log(num); // 1 2 3 5 8...2584 4181 6765
}
```

### 46.5.2 비동기 처리
제너레이터 함수는 next와 yield를 사용하여 호출자와 함수 간 상태를 주고 받을 수 있음
이러한 특성을 이용해 비동기 처리를 동기처럼 보이도록 구현할 수 있음
즉, 프로미스의  `then` /`catch`/`finally` 없이 비동기 처리 결과를 반환하도록 구현할 수 있음
```javascript
// node-fetch는 node.js 환경에서 window.fetch 함수를 사용하기 위한 패키지다.
// 브라우저 환경에서 이 예제를 실행한다면 아래 코드는 필요 없다.
// https://github.com/node-fetch/node-fetch
const fetch = require('node-fetch');

// 제너레이터 실행기
const async = generatorFunc => {
  const generator = generatorFunc(); // ②

  const onResolved = arg => {
	  const result = generator.next(arg); // ⑤ result.value => 프로미스

    return result.done   // false면 재귀
      ? result.value // (10) 함수 종료
      : result.value.then(res => onResolved(res)); // ⑦, ⑨ yield 표현식의 결과값을 할당( 실제 할당되는 값은 promise로 감싸진 값은 아님, 하지만 onResolved 자체의 반환값은 결국 then의 반환값이 promise 이기 때문에 promise가 됨 (재귀))
  };

  return onResolved; // ③
};

(async(function* fetchTodo() { // ①
  const url = 'https://jsonplaceholder.typicode.com/todos/1';

  const response = yield fetch(url); // ⑥  프로미스 반환
  const todo = yield response.json(); // ⑧  프로미스 반환
  console.log(todo);
  // {userId: 1, id: 1, title: 'delectus aut autem', completed: false}
})()); // ④, (11) undefined를 감싼 프로미스 반환
```
1. async 함수에 제너레이터 함수(fetchTodo)를 인자로 넣어 호출
2. generator에 fetchTodo 제너레이터 객체를 생성하여 할당
3. generator를 기억하는 onResolved 클로저를 반환
4. 반환된 onResolved 클로저를 바로 실행
5. `generator.next`를 호출하여 제너레이터 함수를 실행한다.
6. `yield fetch(url);` 을 만나서 일시 중지한다. 이터레이터 리절트 객체의 `value` 에  `fetch(url)`가 평가된 값인 `Promise` 객체가 담기게 되고, `done`은 false 가 담겨 반환되고 `result`에 담긴다.
7. result.done이 false 로 평가되어, `result.value.then(res => onResolved(res))` 이 실행되고, result.value 의 프로미스가 resolve 되면, `res -> onResolved(res)` 가 재귀적으로 호출되고, 다음 next를 실행하게 된다. 이때 resolve 된 값이 next의 인자로 전달되므로, `response` 에 담기게 된다.
8. `yield response.json()` 을 만나게 되고 일시정지가 되면서,  `response.json()` 는 프로미스 객체를 반환하고, result에 해당 프로미스 객체를 value로 가지는 이터레이터 리절트 객체를 담게 된다. 이때 done은 여전히 false 이다. 
> response.json()은 프로미스를 반환한다. [Response.json() - Web APIs | MDN](https://developer.mozilla.org/en-US/docs/Web/API/Response/json) 
9. 7과 동일하게 동작하고, resolve된 값인 res가 `todo`에 담기게 된다.
10. todo를 콘솔에 출력하고 제너레이터 함수가 종료되면, {value: undefined, done: true} 인 이터레이터 리절트 객체가 반환되어 result 에 담기게 되고 result.done이 true로 평가되어 result.value인 undefined가 반환된다.
11. 재귀적으로 호출되었던 onResolved를 벗어나면서 프로미스(then()은 내부적으로 반환 값을 Promise.resolve() 로 감싸서 리턴함)에 감싸지면서 반환된다. 즉, undefined가 resolve 되는 Promise가 반환된다.
> *Promise는 결과적으로 여러번 중첩이 발생수 없다.*
> Promise가 중첩되지 못하게 하는 역할은 Promise.resolve 함수가 한다. resolve 함수의 스펙을 보면 resolve 함수는 받은 인자가 Promise이고 같은 생성자를 가진다면, 받은 인자 그 자체를 리턴한다고 되어 있다. hen 함수는 내부적으로 실행된 결괏값을 이용하여 resolve 함수를 호출한 뒤 그 결괏값인 Promise를 돌려준다.
> [ECMAScript6 성공적인 Promise는 중첩되지 않는다.](https://blog.seulgi.kim/2015/09/ecmascript-6-resolved-promise-cannot-be-nested.html)

> 위 예제는 간략하게 이해를 돕기 위한 함수 이기떄문에 완전하지 않음에 주의하자.
> async / await 을 사용하면, 위 예제의 제너레이터 실행기가 필요없지만, 필요 하다면 직접 구현보다는 co 라이브러리를 사용하는 것이 좋다.
``` javascript
const fetch = require('node-fetch');
// https://github.com/tj/co
const co = require('co');

co(function* fetchTodo() {
  const url = 'https://jsonplaceholder.typicode.com/todos/1';

  const response = yield fetch(url);
  const todo = yield response.json();
  console.log(todo);
  // { userId: 1, id: 1, title: 'delectus aut autem', completed: false }
});
```
## 46.6 async / await
제너레이터를 사용해서 비동기 처리를 동기 처리처럼 동작하도록 구현했지만 코드가 무척
이나 장황해지고 가독성도 좋지 않음
ESS(ECMAScript 2017)에서는 제너레이터보다 간단하고 가독성 좋게 비동기 처리를 동 기 처리처럼 동작하도록 구현할 수 있는 `async/await`가 도입됨
`async/await`는 프로미스를 기반으로 동작
async/await를 사용하면 프로미스의 `then/catch/finally` 후속 처리 메서드에 콜백 함수를 해서 비동기 처리 결과를 후속 처리할 필요 없이 마치 *동기 처리*처럼 프 로미스를 사용할 수 있음
```javascript
const fetch = require('node-fetch');

async function fetchTodo() {
  const url = 'https://jsonplaceholder.typicode.com/todos/1';

  const response = await fetch(url);
  const todo = await response.json();
  console.log(todo);
  // {userId: 1, id: 1, title: 'delectus aut autem', completed: false}
}

fetchTodo();
```
### 46.6.1 async 함수
await 키워드는 반드시 async 함수 내부에서 사용할 수 있음
async 함수는 언제나 프로미스를 반환한다.
명시적으로 프로미스를 반환하지 않아도, async 함수는 암묵적으로 반환값을 resolve하는 프로미스를 반환함
	* 내부에서 await을 사용하지 않아도, Promise에 감싸져서 반환됨
	* 코드는 처음 await을 만날때까지 실행이 지속되고, await을 만나는 순간 프로미스가 반환됨 (마치 제너레이터 실행기의 yield 처럼) 
```javascript
// async 함수 선언문
async function foo(n) { return n; }
foo(1).then(v => console.log(v)); // 1

// async 함수 표현식
const bar = async function (n) { return n; };
bar(2).then(v => console.log(v)); // 2

// async 화살표 함수
const baz = async n => n;
baz(3).then(v => console.log(v)); // 3

// async 메서드
const obj = {
  async foo(n) { return n; }
};
obj.foo(4).then(v => console.log(v)); // 4

// async 클래스 메서드
class MyClass {
  async bar(n) { return n; }
}
const myClass = new MyClass();
myClass.bar(5).then(v => console.log(v)); // 5
```
클래스의 constructor 메서드는 async 메서드가 될 수 없음 (constructor 메서드는 프로미스가 아닌 인스턴스를 반환해야 하기 때문)
```javascript
class MyClass {
  async constructor() { }
  // SyntaxError: Class constructor may not be an async method
}

const myClass = new MyClass();
```
### 46.6.2 await 키워드
await 키워드는 프로미스가 `settled` 상태(비동기 처리가 수행된 상태)가 될 때까지 대기하다가 `settled` 상태가 되면 프로미스가 `resolve`한 처리 결과를 반환함
	* await 키워드는 반드시 프로미스 앞에서 사용해야 함
```javascript
const fetch = require('node-fetch');

const getGithubUserName = async id => {
  const res = await fetch(`https://api.github.com/users/${id}`); // ①
  const { name } = await res.json(); // ②
  console.log(name); // Ungmo Lee
};

getGithubUserName('ungmo2');
```
1. fetch 함수가 수행한 HTTP 요청에 대한 서버의 응답이 도착해서 fetch 함수가 반환한 프로미스가 `settled` 상태가 될 때까지 대기
2. 프로미스가 `settled` 상태가 되면 프로미스가 resolve한 처리 결과가 res 변수에 할당 

await 키워드는 실행을 일시중지시켰다가 프로미스가 settled상태가 되면 다시 재개함 
```javascript
async function foo() {
  const a = await new Promise(resolve => setTimeout(() => resolve(1), 3000));
  const b = await new Promise(resolve => setTimeout(() => resolve(2), 2000));
  const c = await new Promise(resolve => setTimeout(() => resolve(3), 1000));

  console.log([a, b, c]); // [1, 2, 3]
}

foo(); // 약 6초 소요된다.
```

모든 Promise에 async / await 을 사용하는 것이 좋은 것은 아님 
	* 동기적으로(순서대로) 작업이 필요한 경우에는 async / await 이 좋은 선택이 될 수 있지만 비동기적으로(순서없이 동시에) 작업이 필요한 경우에는 오히려 독이된다.
	* 순서에 상관없이 동시다발적으로 필요한 작업에는 `Promise.all` 을 쓰는 것이 좋다.
```javascript
async function foo() {
  const res = await Promise.all([
    new Promise(resolve => setTimeout(() => resolve(1), 3000)),
    new Promise(resolve => setTimeout(() => resolve(2), 2000)),
    new Promise(resolve => setTimeout(() => resolve(3), 1000))
  ]);

  console.log(res); // [1, 2, 3]
}

foo(); // 약 3초 소요된다.

```
```javascript
async function bar(n) {
  const a = await new Promise(resolve => setTimeout(() => resolve(n), 3000));
  // 두 번째 비동기 처리를 수행하려면 첫 번째 비동기 처리 결과가 필요하다.
  const b = await new Promise(resolve => setTimeout(() => resolve(a + 1), 2000));
  // 세 번째 비동기 처리를 수행하려면 두 번째 비동기 처리 결과가 필요하다.
  const c = await new Promise(resolve => setTimeout(() => resolve(b + 1), 1000));

  console.log([a, b, c]); // [1, 2, 3]
}

bar(1); // 약 6초 소요된다.
```
```javascript
async function foo() {
  const res = [];
  const promises = [
    new Promise((resolve) => setTimeout(() => resolve(1), 3000)),
    new Promise((resolve) => setTimeout(() => resolve(2), 2000)),
    new Promise((resolve) => setTimeout(() => resolve(3), 1000)),
  ];
  for (let p of promises) {
    res.push(await p);
  }

  console.log(res); // [1, 2, 3]
}

foo();
```

### 46.6.3 에러 처리
비동기 처리를 위한 콜백 패턴에서 에러처리는 가장 어려운 부분 중 하나임
	* 에러는 호출자 방향으로 전파(콜 스택의 아래 방향)되지만, 비동기 함수의 콜백 함수를 호출한 것은 비동기 함수가 아니기 떄문에 try ... catch 문을 사용해서 에러를 캐치할 수 없음
```javascript
try {
  setTimeout(() => { throw new Error('Error!'); }, 1000);
} catch (e) {
  // 에러를 캐치하지 못한다
  console.error('캐치한 에러', e);
}
```

async/await에서 에러 처리는 try... catch 문을 사용할 수 있음
콜백 함수를 인수로 전달 받는 비동기 함수 함수와는 달리 프로미스를 반환하는 비동기 함수는 명시적 으로 호출할수 있기 때문에 호출자가 명확함

```javascript
const fetch = require('node-fetch');

const foo = async () => {
  try {
    const wrongUrl = 'https://wrong.url';

    const response = await fetch(wrongUrl);
    const data = await response.json();
    console.log(data);
  } catch (err) {
    console.error(err); // TypeError: Failed to fetch
  }
};

foo();
```

	* 위 예제의 foo함수의 catch문은 HTTP 통신에서 발생한 네트워크 에러뿐 아니라 try코드블록내의 모든 문에서 발생한 일반적인 에러까지 모두 캐치할 수 있음
	* async함수내에서 catch문을 사용해서 에러처리를 하지않으면 async 함수는 발생한 에러를 reject하는 프로미스를 반환함
	* async 함수를 호출하고 Promise.prototype.catch 후속처리 메서드를 사용해 에러를 캐치할 수도 있음 
```javascript
const fetch = require('node-fetch');

const foo = async () => {
  const wrongUrl = 'https://wrong.url';

  const response = await fetch(wrongUrl);
  const data = await response.json();
  return data;
};

foo()
  .then(console.log)
  .catch(console.error); // TypeError: Failed to fetch
```

#Javascript_deepdive